client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java

```diff
@@ -522,11 +529,216 @@ protected void set(K key, V value) {
+      }
+
+      @Override
+      public void clear() {
+         RemoteCacheImpl.this.clear();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+         return RemoteCacheImpl.this.containsKey(o);
+      }
+
+      @Override
+      public boolean remove(Object o) {
+         return RemoteCacheImpl.this.remove(o) != null;
+      }
+
+      @Override
+      public boolean removeAll(Collection<?> c) {
+         boolean removedSomething = false;
+         for (Object key : c) {
+            removedSomething |= remove(key);
+         }
+         return removedSomething;
+      }
+   }
+
+   private CloseableIterator<Entry<K, VersionedValue<V>>> castIterator(CloseableIterator iterator) {
+      return iterator;
+   }
+
+   private boolean removeEntry(Map.Entry<K, V> entry) {
+      K key = entry.getKey();
+      VersionedValue<V> versionedValue = RemoteCacheImpl.this.getVersioned(key);
+      if (versionedValue != null && entry.getValue().equals(versionedValue.getValue())) {
+         return RemoteCacheImpl.this.removeWithVersion(key, versionedValue.getVersion());
+      }
+      return false;
+   }
+
+   private class EntrySet extends AbstractCollection<Map.Entry<K, V>> implements CloseableIteratorSet<Entry<K, V>> {
+
+      @Override
+      public CloseableIterator<Entry<K, V>> iterator() {
+         return new CloseableIteratorMapper<>(new RemovableCloseableIterator<>(
+               // First make <K, VersionedValue<V>>
+               castIterator(retrieveEntriesWithMetadata(null, batchSize)),
+               // Apply remove using version
+               e -> removeWithVersion(e.getKey(), e.getValue().getVersion())),
+               // Convert to <K, V> for user
+               e -> new AbstractMap.SimpleImmutableEntry<>(e.getKey(), e.getValue().getValue()));
+      }
+
+      @Override
+      public CloseableSpliterator<Entry<K, V>> spliterator() {
+         return Closeables.spliterator(iterator(), Long.MAX_VALUE, Spliterator.NONNULL | Spliterator.CONCURRENT);
+      }
+
+      @Override
+      public Stream<Entry<K, V>> stream() {
+         return Closeables.stream(spliterator(), false);
+      }
+
+      @Override
+      public Stream<Entry<K, V>> parallelStream() {
+         return Closeables.stream(spliterator(), true);
+      }
+
+      @Override
+      public int size() {
+         return RemoteCacheImpl.this.size();
+      }
+
+      @Override
+      public void clear() {
+         RemoteCacheImpl.this.clear();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+         Map.Entry entry = toEntry(o);
+         if (entry != null) {
+            V value = RemoteCacheImpl.this.get(entry.getKey());
+            return value != null && value.equals(entry.getValue());
+         }
+         return false;
+      }
+
+      @Override
+      public boolean remove(Object o) {
+         Map.Entry<K, V> entry = toEntry(o);
+         return entry != null && RemoteCacheImpl.this.removeEntry(entry);
```

Comment:
`RemoteCacheSupport` contains `remove(K key, V value)` which throws `UnsupportedOperationException`. I think either this method should throw that, too (and then `removeEntry` could be deleted) or plain `RemoteCache.remove(K, V)` should be supported, too.